### PR TITLE
Add only GPU and only CPU version for IVF_SQ8 and IVF_FLAT

### PR DESCRIPTION
**Pull request by fishpenguin**

### CODE REVIEW

**Review by cydrain — APPROVED** [review]

File: core/src/scheduler/optimizer/OnlyCPUPass.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    auto gpu_id = get_gpu_pool();
+    if (gpu_id.empty()) {
+        ResourcePtr res_ptr = ResMgrInst::GetInstance()->GetResource("cpu");
+        auto label = std::make_shared<SpecResLabel>(std::weak_ptr<Resource>(res_ptr));
+        task->label() = label;
+        return true;
+    }
+    return false;
```
Comment (scsven) [inline core/src/scheduler/optimizer/OnlyCPUPass.cpp]: "if (not gpu_id.empty()) return;"

return true only at end of Run()

File: core/src/scheduler/optimizer/OnlyGPUPass.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    server::Config& config = server::Config::GetInstance();
+    std::vector<std::string> search_resources;
+    config.GetResourceConfigSearchResources(search_resources);
+    for (auto& resource : search_resources) {
+        if (resource == "cpu") {
+            return false;
+        }
+    }
```
Comment (scsven) [inline core/src/scheduler/optimizer/OnlyGPUPass.cpp]: ResMgrInst::GetInstance->GetCpuResources()->size() == 0 && GetGpuResources() > 0

File: core/src/scheduler/optimizer/OnlyGPUPass.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        task->label() = label;
+    } else {
+        return false;
+    }
+
+    if (specified_gpu_id_ < gpu_id.size() - 1) {
+        ++specified_gpu_id_;
+    } else {
+        specified_gpu_id_ = 0;
+    }
```
Comment (scsven) [inline core/src/scheduler/optimizer/OnlyGPUPass.cpp]: specified_gpu_id = specified_gpu_id++ % gpu_id.size()

File: core/src/scheduler/optimizer/OnlyGPUPass.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    auto gpu_id = get_gpu_pool();
+    if (!gpu_id.empty()) {
+        ResourcePtr res_ptr = ResMgrInst::GetInstance()->GetResource(ResourceType::GPU, gpu_id[specified_gpu_id_]);
+        auto label = std::make_shared<SpecResLabel>(std::weak_ptr<Resource>(res_ptr));
+        task->label() = label;
+    } else {
+        return false;
+    }
```
Comment (scsven) [inline core/src/scheduler/optimizer/OnlyGPUPass.cpp]: if (gpu_id.empty()) return false;

**Review by wxyucs — CHANGES_REQUESTED** [review]

File: core/src/scheduler/optimizer/OnlyGPUPass.cpp (hunk 2 above)
Comment (scsven) [reply core/src/scheduler/optimizer/OnlyGPUPass.cpp]: Read it from config.

**Review by wxyucs — APPROVED** [review]

**Pull request merged**